src/lib/openssl_preload/crypto.c

```diff
@@ -0,0 +1,219 @@
+/*
+ * The Shadow Simulator
+ * Copyright (c) 2010-2011, Rob Jansen
+ * See LICENSE for licensing information
+ */
+
+#define _GNU_SOURCE
+#include <dlfcn.h>
+#include <execinfo.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+// Caches whether or not a memory address is from libssl.so
+typedef struct {
+    void* addr;
+    bool is_libssl;
+} bt_cache_entry_t;
+
+// We use a small cache size: in ad-hoc experiments with tor-0.4.6.9, we observed
+// at most three callers of EVP_EncryptUpdate.
+#define EVP_BACKTRACE_CACHE_LEN 10
+static bt_cache_entry_t evp_backtrace_cache[EVP_BACKTRACE_CACHE_LEN];
```

Comment:
This cache doesn't appear to be thread safe; it probably ought to be a protected with e.g. a `pthread_mutex_t`.

Technically we shouldn't have true parallel access if it's running under Shadow, as long as no blocking syscalls are made in the middle of accessing it. (And we don't introduce true preemption in Shadow in the future). I'd lean towards just putting in and using a mutex to be  safe, but I'd be ok with just documenting that this code isn't thread-safe except when running under Shadow.

Reply:
Added a global mutex to protect global state.